torch/distributed/distributed_c10d.py

```diff
@@ -2082,6 +2082,11 @@ def all_gather(tensor_list, tensor, group=None, async_op=False):
         [tensor([1.+1.j, 2.+2.j]), tensor([3.+3.j, 4.+4.j])] # Rank 1
 
     """
+    # `tensor_list` would be understood as "tensor or list."
+    # If it is a single tensor, we would route it to the _all_gather_base implementation.
+    if isinstance(tensor_list, torch.Tensor):
+        return _all_gather_base(tensor_list, tensor, group, async_op)
```

Comment:
how about a new flag into `dist.all_gather` such as `inplace=True`? And when this is provided, we document that the first arg should be an appropriately sized tensor and we fill the result in that?

Reply:
This PR is tangential to in-place vs out-of-place. It is more about in which format the user wants to receive the all-gather result.